url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/io/LinkedBufferInput.java

```diff
@@ -0,0 +1,399 @@
+//
+// MessagePack for Java
+//
+// Copyright (C) 2009 - 2013 FURUHASHI Sadayuki
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+package org.msgpack.io;
+
+import java.io.IOException;
+import java.io.EOFException;
+import java.util.LinkedList;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.nio.ByteBuffer;
+
+public class LinkedBufferInput extends AbstractInput {
+    LinkedList<ByteBuffer> link;
+
+    int writable;
+
+    private int nextAdvance;
+
+    private byte[] tmpBuffer;
+
+    private ByteBuffer tmpByteBuffer;
+
+    private final int bufferSize;
+
+    public LinkedBufferInput(int bufferSize) {
+        this.link = new LinkedList<ByteBuffer>();
+        this.writable = -1;
+        this.tmpBuffer = new byte[8];
+        this.tmpByteBuffer = ByteBuffer.wrap(tmpBuffer);
+        this.bufferSize = bufferSize;
+    }
+
+    public int read(byte[] b, int off, int len) throws EOFException {
+        if (link.isEmpty()) {
+            return 0;
+        }
+        int olen = len;
+        while (true) {
+            ByteBuffer bb = link.getFirst();
+            if (len < bb.remaining()) {
+                bb.get(b, off, len);
+                incrReadByteCount(len);
+                return olen;
+            }
+            int rem = bb.remaining();
+            bb.get(b, off, rem);
+            incrReadByteCount(rem);
+            len -= rem;
+            off += rem;
+            if (!removeFirstLink(bb)) {
+                break;
+            }
+        }
+        return olen - len;
+    }
+
+    public boolean tryRefer(BufferReferer ref, int len) throws IOException {
+        ByteBuffer bb = null;
+        try {
+            bb = link.getFirst();
+        } catch(NoSuchElementException e) {}
+        if (bb == null) {
+            throw new EndOfBufferException();
+        } else if (bb.remaining() < len) {
+            return false;
+        }
+        boolean success = false;
+        int pos = bb.position();
+        int lim = bb.limit();
+        try {
+            bb.limit(pos + len);
+            ref.refer(bb, true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;ref&#39; (index: 0 | type: BufferReferer). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;len&#39;. The same parameter type is validated here for example: url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/unpacker/MessagePackUnpacker.java:358. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).